### PR TITLE
Add single simultaneous replica movement strategy

### PR DESCRIPTION
**Pull request by leizor**
This PR adds `SingleReplicaMovementStrategy` which operates like `BaseReplicaMovementStrategy` except that it expands any execution tasks that contain multiple replica movements into multiple executions tasks that contain single replica movements.

### CODE REVIEW

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/detector/StatefulSetBrokerProvisioner.java [hunk 1]
```diff
@@ -42,7 +38,7 @@ public void configure(Map<String, ?> configs) {
     @Override
     protected ProvisionerState addOrRemoveBrokers(ProvisionRecommendation rec) {
 
-        try (final KubernetesClient client = new KubernetesClientBuilder().build()) {
+        try (KubernetesClient client = new KubernetesClientBuilder().build()) {
```
Comment (replay) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/detector/StatefulSetBrokerProvisioner.java:41]: why should `KubernetesClient` not be final?

File: cruise-control/src/test/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/prometheus/PrometheusMetricSamplerTest.java [hunk 2]
```diff
@@ -80,22 +80,6 @@ public void setUp() {
...
-    @Test(expected = ConfigException.class)
-    public void testConfigureWithPrometheusEndpointNoPortFails() throws Exception {
-        Map<String, Object> config = new HashMap<>();
-        config.put(PROMETHEUS_SERVER_ENDPOINT_CONFIG, "http://kafka-cluster-1.org");
-        addCapacityConfig(config);
-        _prometheusMetricSampler.configure(config);
-    }
-
-    @Test(expected = ConfigException.class)
-    public void testConfigureWithPrometheusEndpointNegativePortFails() throws Exception {
```
Comment (replay) [inline cruise-control/src/test/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/prometheus/PrometheusMetricSamplerTest.java:92]: the names of these two tests don't sound like they're related to what this PR is doing, what's the reason to remove them?

**Review by replay — APPROVED** [review]
This looks good to me, thx.

Since this expands single tasks which move multiple partition replicas into multiple tasks, to ensure that we never add multiple followers to one leader at the same time we also need to configure the task concurrency settings in such a way that don't allow that two of the expanded tasks get executed concurrently. Do we have a setting for that?

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/detector/StatefulSetBrokerProvisioner.java (hunk 1 above)
Comment (leizor) [reply cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/detector/StatefulSetBrokerProvisioner.java:41]: I think it should be, but the Cruise Control repo has some pretty opinionated lints and this was one of them.

File: cruise-control/src/test/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/prometheus/PrometheusMetricSamplerTest.java (hunk 2 above)
Comment (leizor) [reply cruise-control/src/test/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/prometheus/PrometheusMetricSamplerTest.java:92]: https://github.com/grafana/cruise-control/commit/ae04ed7a54bb3cda17350c307ebd648e96ed786a actually broke these tests (negative/no explicit port should be valid) but I hadn't been running repo-wide tests till now so I never noticed; this was just to get all tests passing.

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/detector/StatefulSetBrokerProvisioner.java (hunk 1 above)
Comment (leizor) [reply cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/detector/StatefulSetBrokerProvisioner.java:41]: (In general, they don't use `final` in a lot of places that other Java developers might...but I think the "correct" use of `final` is one of those tabs vs spaces type discussions in the Java world)

**Comment by leizor** [comment]
> Since this expands single tasks which move multiple partition replicas into multiple tasks, to ensure that we never add multiple followers to one leader at the same time we also need to configure the task concurrency settings in such a way that don't allow that two of the expanded tasks get executed concurrently. Do we have a setting for that?

Yep! Now that we have each replica movement broken out into its own execution task, the same `num.concurrent.partition.movements.per.broker` config we already have set to `1` will serve to enforce that.

**Pull request merged**